File: python/230419.py
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        pathLength = 0
        
        def dfs(root : Optional[TreeNode], left, right):
            if root:
                nonlocal pathLength
                pathLength = max(pathLength, max(left, right))
                dfs(root.left, right + 1, 0)
                dfs(root.right, 0, left + 1)

        dfs(root, 0, 0)
        return pathLength

File: dart/lib/leetcode/2304/230419.dart
import 'package:dart/leetcode/helpers/treenode.dart';

import 'dart:math';
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */

class Solution {
  int longestZigZag(TreeNode? root) {
    int ret = 0;

    void f(TreeNode? root, int left, int right) {
      if (root != null) {
        ret = max(ret, max(left, right));
        f(root.left, right + 1, 0);
        f(root.right, 0, left + 1);
      }
    }

    f(root, 0, 0);
    return ret;
  }
}
